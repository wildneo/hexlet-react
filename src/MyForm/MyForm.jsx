/* eslint-disable jsx-a11y/no-onchange */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';

export default class MyForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      form: {
        email: '',
        password: '',
        address: '',
        city: '',
        country: '',
        acceptRules: false,
      },
      formState: 'form',
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const { form } = this.state;
    const { target, target: { name } } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    this.setState({ form: { ...form, [name]: value } });
  }

  handleSubmit(event) {
    event.preventDefault();
    this.setState({ formState: 'table' });
  }

  handleClick(event) {
    event.preventDefault();
    this.setState({ formState: 'form' });
  }

  renderForm() {
    const { form } = this.state;

    return (
      <form onSubmit={this.handleSubmit}>
        <div className="form-row">
          <div className="form-group col-md-6">
            <label htmlFor="inputEmail4" className="col-form-label">Email</label>
            <input
              className="form-control"
              id="inputEmail4"
              name="email"
              type="email"
              placeholder="Email"
              onChange={this.handleChange}
              value={form.email}
            />
          </div>
          <div className="form-group col-md-6">
            <label htmlFor="inputPassword4" className="col-form-label">Password</label>
            <input
              className="form-control"
              id="inputPassword4"
              name="password"
              type="password"
              placeholder="Password"
              onChange={this.handleChange}
              value={form.password}
            />
          </div>
        </div>
        <div className="form-group">
          <label htmlFor="inputAddress" className="col-form-label">Address</label>
          <textarea
            className="form-control"
            id="inputAddress"
            name="address"
            type="text"
            placeholder="1234 Main St"
            onChange={this.handleChange}
            value={form.address}
          />
        </div>
        <div className="form-row">
          <div className="form-group col-md-6">
            <label htmlFor="inputCity" className="col-form-label">City</label>
            <input
              className="form-control"
              id="inputCity"
              name="city"
              type="text"
              onChange={this.handleChange}
              value={form.city}
            />
          </div>
          <div className="form-group col-md-6">
            <label htmlFor="inputCountry" className="col-form-label">Country</label>
            <select
              id="inputCountry"
              className="form-control"
              name="country"
              onChange={this.handleChange}
              value={form.country}
            >
              <option>Choose</option>
              <option value="argentina">Argentina</option>
              <option value="russia">Russia</option>
              <option value="china">China</option>
            </select>
          </div>
        </div>
        <div className="form-group">
          <div className="form-check">
            <label className="form-check-label" htmlFor="rules">
              <input
                id="rules"
                className="form-check-input"
                name="acceptRules"
                type="checkbox"
                onChange={this.handleChange}
                checked={form.acceptRules}
              />
              Accept Rules
            </label>
          </div>
        </div>
        <button type="submit" className="btn btn-primary">Sign in</button>
      </form>
    );
  }

  renderTable() {
    const { form } = this.state;
    const data = Object.entries(form);
    const row = ([key, value]) => (
      <tr key={key}>
        <td>{key}</td>
        <td>{value.toString()}</td>
      </tr>
    );

    return (
      <div>
        <button type="button" onClick={this.handleClick}>Back</button>
        <table className="table">
          <tbody>
            {data.sort().map(row)}
          </tbody>
        </table>
      </div>
    );
  }

  render() {
    const { formState } = this.state;
    switch (formState) {
      case 'form':
        return this.renderForm();

      case 'table':
        return this.renderTable();

      default:
        throw new Error(`Unexpected state ${formState}`);
    }
  }
}
