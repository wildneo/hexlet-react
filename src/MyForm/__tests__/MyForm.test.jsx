import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MyForm from '../MyForm';

Enzyme.configure({ adapter: new Adapter() });

test('MyForm', async () => {
  const wrapper = mount(<MyForm />);
  expect(wrapper.render()).toMatchSnapshot();

  const input = wrapper.find('form');
  input.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const button = wrapper.find('button');
  button.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const inputAcceptRules = wrapper.find('input[name="acceptRules"]');
  inputAcceptRules.simulate('change', { target: { name: 'acceptRules', type: 'checkbox', checked: true } });

  const inputCity = wrapper.find('input[name="city"]');
  inputCity.simulate('change', { target: { type: 'input', name: 'city', value: 'MyCity' } });

  const inputEmail = wrapper.find('input[name="email"]');
  inputEmail.simulate('change', { target: { type: 'input', name: 'email', value: 'test@email.com' } });

  const inputPassword = wrapper.find('input[name="password"]');
  inputPassword.simulate('change', { target: { type: 'input', name: 'password', value: 'mysuperpass' } });

  const input2 = wrapper.find('form');
  input2.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const button2 = wrapper.find('button');
  button2.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
