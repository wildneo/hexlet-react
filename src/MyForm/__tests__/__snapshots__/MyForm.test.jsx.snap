// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyForm 1`] = `
<form>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputEmail4"
      >
        Email
      </label>
      <input
        class="form-control"
        id="inputEmail4"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputPassword4"
      >
        Password
      </label>
      <input
        class="form-control"
        id="inputPassword4"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="col-form-label"
      for="inputAddress"
    >
      Address
    </label>
    <textarea
      class="form-control"
      id="inputAddress"
      name="address"
      placeholder="1234 Main St"
      type="text"
    />
  </div>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCity"
      >
        City
      </label>
      <input
        class="form-control"
        id="inputCity"
        name="city"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCountry"
      >
        Country
      </label>
      <select
        class="form-control"
        id="inputCountry"
        name="country"
      >
        <option>
          Choose
        </option>
        <option
          value="argentina"
        >
          Argentina
        </option>
        <option
          value="russia"
        >
          Russia
        </option>
        <option
          value="china"
        >
          China
        </option>
      </select>
    </div>
  </div>
  <div
    class="form-group"
  >
    <div
      class="form-check"
    >
      <label
        class="form-check-label"
        for="rules"
      >
        <input
          class="form-check-input"
          id="rules"
          name="acceptRules"
          type="checkbox"
        />
        Accept Rules
      </label>
    </div>
  </div>
  <button
    class="btn btn-primary"
    type="submit"
  >
    Sign in
  </button>
</form>
`;

exports[`MyForm 2`] = `
<div>
  <button
    type="button"
  >
    Back
  </button>
  <table
    class="table"
  >
    <tbody>
      <tr>
        <td>
          acceptRules
        </td>
        <td>
          false
        </td>
      </tr>
      <tr>
        <td>
          address
        </td>
        <td />
      </tr>
      <tr>
        <td>
          city
        </td>
        <td />
      </tr>
      <tr>
        <td>
          country
        </td>
        <td />
      </tr>
      <tr>
        <td>
          email
        </td>
        <td />
      </tr>
      <tr>
        <td>
          password
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`MyForm 3`] = `
<form>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputEmail4"
      >
        Email
      </label>
      <input
        class="form-control"
        id="inputEmail4"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputPassword4"
      >
        Password
      </label>
      <input
        class="form-control"
        id="inputPassword4"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="col-form-label"
      for="inputAddress"
    >
      Address
    </label>
    <textarea
      class="form-control"
      id="inputAddress"
      name="address"
      placeholder="1234 Main St"
      type="text"
    />
  </div>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCity"
      >
        City
      </label>
      <input
        class="form-control"
        id="inputCity"
        name="city"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCountry"
      >
        Country
      </label>
      <select
        class="form-control"
        id="inputCountry"
        name="country"
      >
        <option>
          Choose
        </option>
        <option
          value="argentina"
        >
          Argentina
        </option>
        <option
          value="russia"
        >
          Russia
        </option>
        <option
          value="china"
        >
          China
        </option>
      </select>
    </div>
  </div>
  <div
    class="form-group"
  >
    <div
      class="form-check"
    >
      <label
        class="form-check-label"
        for="rules"
      >
        <input
          class="form-check-input"
          id="rules"
          name="acceptRules"
          type="checkbox"
        />
        Accept Rules
      </label>
    </div>
  </div>
  <button
    class="btn btn-primary"
    type="submit"
  >
    Sign in
  </button>
</form>
`;

exports[`MyForm 4`] = `
<div>
  <button
    type="button"
  >
    Back
  </button>
  <table
    class="table"
  >
    <tbody>
      <tr>
        <td>
          acceptRules
        </td>
        <td>
          true
        </td>
      </tr>
      <tr>
        <td>
          address
        </td>
        <td />
      </tr>
      <tr>
        <td>
          city
        </td>
        <td>
          MyCity
        </td>
      </tr>
      <tr>
        <td>
          country
        </td>
        <td />
      </tr>
      <tr>
        <td>
          email
        </td>
        <td>
          test@email.com
        </td>
      </tr>
      <tr>
        <td>
          password
        </td>
        <td>
          mysuperpass
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`MyForm 5`] = `
<form>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputEmail4"
      >
        Email
      </label>
      <input
        class="form-control"
        id="inputEmail4"
        name="email"
        placeholder="Email"
        type="email"
        value="test@email.com"
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputPassword4"
      >
        Password
      </label>
      <input
        class="form-control"
        id="inputPassword4"
        name="password"
        placeholder="Password"
        type="password"
        value="mysuperpass"
      />
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="col-form-label"
      for="inputAddress"
    >
      Address
    </label>
    <textarea
      class="form-control"
      id="inputAddress"
      name="address"
      placeholder="1234 Main St"
      type="text"
    />
  </div>
  <div
    class="form-row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCity"
      >
        City
      </label>
      <input
        class="form-control"
        id="inputCity"
        name="city"
        type="text"
        value="MyCity"
      />
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        class="col-form-label"
        for="inputCountry"
      >
        Country
      </label>
      <select
        class="form-control"
        id="inputCountry"
        name="country"
      >
        <option>
          Choose
        </option>
        <option
          value="argentina"
        >
          Argentina
        </option>
        <option
          value="russia"
        >
          Russia
        </option>
        <option
          value="china"
        >
          China
        </option>
      </select>
    </div>
  </div>
  <div
    class="form-group"
  >
    <div
      class="form-check"
    >
      <label
        class="form-check-label"
        for="rules"
      >
        <input
          checked=""
          class="form-check-input"
          id="rules"
          name="acceptRules"
          type="checkbox"
        />
        Accept Rules
      </label>
    </div>
  </div>
  <button
    class="btn btn-primary"
    type="submit"
  >
    Sign in
  </button>
</form>
`;
